### PR TITLE
Introduced skeleton for a Roslyn-based GenAPI (#10411).

### PR DESCRIPTION
**Pull request by andriipatsula**
Created Microsoft.DotNet.GenAPI.Shared, Microsoft.DotNet.GenAPI.Tasks, Microsoft.DotNet.GenAPI.Tool projects. Added interfaces for assembly loading, filtering and ordering symbol assemblies.

### To double check:

* [ ] The right tests are in and and the right validation has happened.  Guidance:  https://github.com/dotnet/arcade/tree/main/Documentation/Validation

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/arcade?pullRequest=10567) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/arcade?pullRequest=10567) before we can accept your contribution.<br/><br/>:x: andriipatsula [sign now](https://cla.dotnetfoundation.org/dotnet/arcade?pullRequest=10567)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/arcade?pullRequest=10567) it.</sub>

File: src/Microsoft.DotNet.GenAPI/Shared/AssemblySymbolLoader.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+namespace Microsoft.DotNet.GenAPI.Shared
+{
+    internal class AssemblySymbolLoader : IAssemblySymbolLoader
+    {
+        public IAssemblySymbol? LoadAssembly(string path)
+        {
+            using (var stream = File.OpenRead(path))
+            {
+                return LoadAssembly(stream);
+            }
```
Comment (premun) [inline src/Microsoft.DotNet.GenAPI/Shared/AssemblySymbolLoader.cs:18]: FYI, this is also possible:
```suggestion
            using var stream = File.OpenRead(path);
            return LoadAssembly(stream);
```

File: src/Microsoft.DotNet.GenAPI/Shared/IAssemblySymbolOrderProvider.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        /// </summary>
+        /// <param name="namespaces">List of TypeMembers to be sorted.</param>
+        /// <returns>Returns TypeMembers in sorted order.</returns>
+        IEnumerable<T> OrderTypes<T>(IEnumerable<T> symbols) where T : ITypeSymbol;
+        /// <summary>
+        /// Sorts the elements of a ISymbol.
+        /// </summary>
+        /// <param name="namespaces">List of Members to be sorted.</param>
+        /// <returns>Returns Members in sorted order.</returns>
+        IEnumerable<ISymbol> OrderMembers(IEnumerable<ISymbol> members);
```
Comment (premun) [inline src/Microsoft.DotNet.GenAPI/Shared/IAssemblySymbolOrderProvider.cs:31]:
```suggestion
        /// <summary>
        /// Sorts the elements of a INamespaceSymbol.
        /// </summary>
        /// <param name="namespaces">List of namespaces to be sorted.</param>
        /// <returns>Returns namespaces in sorted order.</returns>
        IEnumerable<INamespaceSymbol> OrderNamespaces(IEnumerable<INamespaceSymbol> namespaces);

        /// <summary>
        /// Sorts the elements of a ITypeSymbol.
        /// </summary>
        /// <param name="namespaces">List of TypeMembers to be sorted.</param>
        /// <returns>Returns TypeMembers in sorted order.</returns>
        IEnumerable<T> OrderTypes<T>(IEnumerable<T> symbols) where T : ITypeSymbol;

        /// <summary>
        /// Sorts the elements of a ISymbol.
        /// </summary>
        /// <param name="namespaces">List of Members to be sorted.</param>
        /// <returns>Returns Members in sorted order.</returns>
        IEnumerable<ISymbol> OrderMembers(IEnumerable<ISymbol> members);
```

File: src/Microsoft.DotNet.GenAPI/Tasks/Microsoft.DotNet.GenAPI.Tasks.csproj [hunk 3]
```diff
@@ -0,0 +1,26 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (premun) [inline src/Microsoft.DotNet.GenAPI/Tasks/Microsoft.DotNet.GenAPI.Tasks.csproj:1]: Please rename this + it's folder + the namespace to `Tasks`

**Review by premun — APPROVED** [review]
I am thinking whether a short README in the main GenAPI folder wouldn't do a lot of good explaining the 4 projects and that we have a v2 now?

File: src/Microsoft.DotNet.GenAPI/README.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+Supported parameters in GenAPIv2:
+* **Assembly** - [required] Path for an specific assembly or a directory to get all assemblies.
+* **GenAPILibPath** - Delimited (',' or ';') set of paths to use for resolving assembly references.
+* **ApiList** - Specify a api list in the DocId format of which APIs to include.
+* **ExcludeApiList** - Specify a api list in the DocId format of which APIs to exclude.
+* **OutputPath** - Output path. Default is the console. Can specify an existing directory as well and then a file will be created for each assembly with the matching name of the assembly
+* **HeaderFile** - Specify a file with an alternate header content to prepend to output.
+* **ExceptionMessage** - Method bodies should throw PlatformNotSupportedException.
+* **ExcludeAttributesList** - Specify a list in the DocId format of which attributes should be excluded from being applied on apis.
+* **FollowTypeForwards** - Resolve type forwards and include its members.
```
Comment (premun) [inline src/Microsoft.DotNet.GenAPI/README.md:21]: Honestly, not sure this is a great idea. The markdown will get quickly out of sync so possibly it's more useful to have these descriptions in the XML doc for every property on the task itself which is technically the source of truth.
I expect the CLI will have its own help string too.

**Review by premun — APPROVED** [review]

File: src/Microsoft.DotNet.GenAPI/Shared/AssemblySymbolLoader.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        {
+            PortableExecutableReference reference;
+
+            using (var memoryStream = new MemoryStream())
+            {
+                stream.CopyTo(memoryStream);
+                memoryStream.Position = 0;
+
+                // MetadataReference.CreateFromStream closes the stream
+                reference = MetadataReference.CreateFromStream(memoryStream);
```
Comment (premun) [inline src/Microsoft.DotNet.GenAPI/Shared/AssemblySymbolLoader.cs:29]: Question - why not just use the original file stream directly? Does the `MetadataReference.CreateFromStream` api work with any stream?

**Review by riarenas — DISMISSED** [review]

File: src/Microsoft.DotNet.GenAPI/Shared/AssemblySymbolLoader.cs (hunk 5 above)
Comment (andriipatsula) [reply src/Microsoft.DotNet.GenAPI/Shared/AssemblySymbolLoader.cs:29]: the reason is that `MetadataReference.CreateFromStream closes the stream` (as in comment). I'll need original stream later in `LoadAssembly`.

**Pull request merged**